### PR TITLE
ARROW-17037: [C++] Split utf8.h to avoid exposing xsimd dependency to third-party code

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

File: cpp/src/arrow/array/array_binary.cc [hunk 1]
```diff
@@ -26,6 +26,7 @@
 #include "arrow/type_traits.h"
 #include "arrow/util/checked_cast.h"
 #include "arrow/util/logging.h"
+#include "arrow/util/utf8.h"
```
Comment (cyb70289) [inline cpp/src/arrow/array/array_binary.cc:29]: Is this header necessary? Looks this file calls `internal::ValidateUTF8`, not `util::ValidateUTF8`.

**Comment by cyb70289** [comment]
Benchmark failure is strange. But looks not related to this PR.

File: cpp/src/arrow/array/array_binary.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/array/array_binary.cc:29]: `ValidateUTF8` is part of `utf8.h`, the inlined version is named `ValidateUTF8Inline`.

Comment (cyb70289) [reply cpp/src/arrow/array/array_binary.cc:29]: `ValidateUTF8` in array/validate.h?
https://github.com/apache/arrow/blob/master/cpp/src/arrow/array/validate.h#L53

Comment (pitrou) [reply cpp/src/arrow/array/array_binary.cc:29]: Ah! Sorry, I might have misread then.

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Comment by pitrou** [comment]
@ursabot please benchmark lang=C++

**Review by cyb70289 — APPROVED** [review]
+1

**Pull request merged**